### PR TITLE
Feature/stud130 stud131 stud172

### PR DESCRIPTION
**Pull request by marksman1298**

### CODE REVIEW

**Comment by nuzurie** [comment]
First, can you please rebase your commit on the target branch?

File: messaging/repository/chat_repository.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	"context"
+	"github.com/gocql/gocql"
+	"time"
+)
+
+const (
+	insertMessage = `INSERT INTO messages (room_id, from_student_id, message_body, sent_timestamp) VALUES (?, ?, ?, ?)`
+	editMessage = `UPDATE messages SET message_body=? WHERE room_id=? AND sent_timestamp=? IF EXISTS;`
+	getMessage = `SELECT * FROM messages where room_id=? AND sent_timestamp =?`
+	getMessages = `SELECT * FROM messages WHERE room_id=? AND sent_timestamp <=? limit ?`
```
Comment (nuzurie) [inline messaging/repository/chat_repository.go]: Does the `<=` also return the last message? For example you extracted the timestamp from the last message you have. Will this query also return that message?

File: messaging/repository/chat_repository.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+func (m *messageRepository) SaveMessage(ctx context.Context, message *domain.Message) error {
+	err := m.dbSession.Query(insertMessage, message.RoomID, message.FromStudentID, message.MessageBody, message.SentTimestamp).WithContext(ctx).Exec()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *messageRepository) EditMessage(ctx context.Context, message *domain.Message) error {
+	var editedMsg domain.Message
```
Comment (nuzurie) [inline messaging/repository/chat_repository.go:39]: why use editedMsg if it isn't used?

File: messaging/repository/chat_repository.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	_, err := m.dbSession.Query(editMessage, message.MessageBody, message.RoomID, message.SentTimestamp).WithContext(ctx).
```
Comment (nuzurie) [inline messaging/repository/chat_repository.go:40]: Do you need to .Exec() after WithContect()?

Comment (nuzurie) [reply messaging/repository/chat_repository.go:40]: or does it work like this?

File: messaging/repository/chat_repository.go [hunk 4]
```diff
@@ -31,13 +35,57 @@ func (m *messageRepository) SaveMessage(ctx context.Context, message *domain.Mes
...
+	var retrievedMessages []domain.Message
+	var scanner gocql.Scanner
+
+	scanner = m.dbSession.Query(getMessages, roomID, timeStamp, limit).WithContext(ctx).Iter().Scanner()
+
+	for scanner.Next(){
+		var msg domain.Message
+		err := scanner.Scan(&msg.RoomID, &msg.SentTimestamp, &msg.FromStudentID, &msg.MessageBody)
+
+		if err == nil {
```
Comment (nuzurie) [inline messaging/repository/chat_repository.go]: log the error. Return the error as internal error, so they can make the call again. after err == nil {} block, append the message

File: messaging/repository/chat_repository.go [hunk 5]
```diff
@@ -31,13 +35,57 @@ func (m *messageRepository) SaveMessage(ctx context.Context, message *domain.Mes
...
+			retrievedMessages = append(retrievedMessages, msg)
+		}
+
+	}
+	if err := scanner.Err(); err != nil {
```
Comment (nuzurie) [inline messaging/repository/chat_repository.go:82]: or does the above error propagate here?

File: messaging/repository/chat_repository.go [hunk 6]
```diff
@@ -31,13 +35,57 @@ func (m *messageRepository) SaveMessage(ctx context.Context, message *domain.Mes
...
+		return nil, err
```
Comment (nuzurie) [inline messaging/repository/chat_repository.go:83]: map all erros in repo to internal errors please

File: messaging/usecase/message_usecase.go [hunk 7]
```diff
@@ -51,14 +52,54 @@ func (u *messageUseCase) SaveMessage(ctx context.Context, message *domain.Messag
...
+	defer cancel()
+
+	//message.SentTimestamp, message.FromStudentID, message.RoomID = timeStamp, userID, roomID
+	//get existing message at timestamp
+	existingMessage, err := u.messageRepository.GetMessage(ctx, roomID, timeStamp)
+	if err != nil {
+		return err
+	}
+
+	//if userID != existingMessage.FromStudentID{
```
Comment (nuzurie) [inline messaging/usecase/message_usecase.go]: why are we not checking this?

File: messaging/usecase/message_usecase.go [hunk 8]
```diff
@@ -51,14 +52,54 @@ func (u *messageUseCase) SaveMessage(ctx context.Context, message *domain.Messag
 	return nil
 }
 
-func (u *messageUseCase) EditMessage(ctx context.Context, userID string, message *domain.Message) error {
-	panic("")
-}
+func (u *messageUseCase) EditMessage(ctx context.Context, roomID string, userID string, timeStamp time.Time, message string) error {
```
Comment (nuzurie) [inline messaging/usecase/message_usecase.go]: can you return the edited message here?

**Review by nuzurie — CHANGES_REQUESTED** [review]

File: messaging/usecase/message_usecase.go [hunk 9]
```diff
@@ -52,33 +52,33 @@ func (u *messageUseCase) SaveMessage(ctx context.Context, message *domain.Messag
...
 	}
 
-	//if userID != existingMessage.FromStudentID{
-	//	return errors.NewUnauthorizedError("Users can only edit their own messages")
-	//}
+	if userID != existingMessage.FromStudentID{
+		return nil, errors.NewUnauthorizedError("Users can only edit their own messages")
+	}
 
 	if message == "" || message == existingMessage.MessageBody {
```
Comment (car12lin12) [inline messaging/usecase/message_usecase.go]: should empty or same body be considered an error? I think if it's empty, then the message should be deleted ( in the future might want to ask the user if they want to delete it first ). And if it's the same, then just return the message

Comment (marksman1298) [reply messaging/usecase/message_usecase.go]: Yeah makes sense. Changed it.

**Comment by nuzurie** [comment]
Tasks:
1. UseCase for messaging including get, edit and delete
2. Repository for messaging
3. Delivery: LoadMessage

**Review by car12lin12 — APPROVED** [review]
Great job Mark!!

**Pull request merged**